scipy/sparse/linalg/_isolve/iterative.py

```diff
@@ -92,28 +92,11 @@ def _get_atol(tol, atol, bnrm2, get_residual, routine_name):
         Name of the routine.
     """
 
-    if atol is None:
-        warnings.warn("scipy.sparse.linalg.{name} called without specifying `atol`. "
-                      "The default value will be changed in a future release. "
-                      "For compatibility, specify a value for `atol` explicitly, e.g., "
-                      "``{name}(..., atol=0)``, or to retain the old behavior "
-                      "``{name}(..., atol='legacy')``".format(name=routine_name),
-                      category=DeprecationWarning, stacklevel=4)
-        atol = 'legacy'
+    if not np.issubdtype(type(atol), np.floating):
+        raise ValueError("Parameter `atol` must be a floating-point number!")
 
     tol = float(tol)
-
-    if atol == 'legacy':
-        # emulate old legacy behavior
-        resid = get_residual()
-        if resid <= tol:
-            return 'exit'
-        if bnrm2 == 0:
-            return tol
-        else:
-            return tol * float(bnrm2)
-    else:
-        return max(float(atol), tol * float(bnrm2))
```

Comment:
@h-vetinari I assume we want to remove this legacy behavior here. This `_get_atol()`function becomes rather small here.

Reply:
Also, about `scipy/sparse/linalg/_isolve/iterative.py:576` 

```
 if callback is not None and callback_type is None:
        # Warn about 'callback_type' semantic changes.
        # Probably should be removed only in far future, Scipy 2.0 or so.
        warnings.warn("scipy.sparse.linalg.gmres called without specifying `callback_type`. "
                      "The default value will be changed in a future release. "
                      "For compatibility, specify a value for `callback_type` explicitly, e.g., "
                      "``{name}(..., callback_type='pr_norm')``, or to retain the old behavior "
                      "``{name}(..., callback_type='legacy')``",
                      category=DeprecationWarning, stacklevel=3)

    if callback_type is None:
        callback_type = 'legacy'
```

Do we also want this warning gone? I suppose its better deprecated now than later.

Reply:
> @h-vetinari I assume we want to remove this legacy behavior here. This `_get_atol()`function becomes rather small after removing the legacy code.

Indeed, that function only exists to work around the deprecation, and we should IMO remove it completely. You'll need to make sure that at each call site, the maximum of atol and the correct(ly calculated) residual is taken.

> Also, about scipy/sparse/linalg/_isolve/iterative.py:576
> [...]
> Do we also want this warning gone? I suppose its better deprecated now than later.

It's deprecated already, and a whole separate issue IMO. I suggest not to touch it in this PR.

Reply:
Alright. I think I'm done with gcrotmk and lgmres and their corresponding test files. However, I've stumbled upon issues when trying to remove the deprecation messages from iterative.py and test_iterative.py because I see some functions receive `solver` as arguments for different functions i.e:

```
  # list of tuples (solver, symmetric, positive_definite )
  solvers = [cg, cgs, bicg, bicgstab, gmres, qmr, minres, lgmres, gcrotmk, tfqmr]
  sym_solvers = [minres, cg]
  posdef_solvers = [cg]
  real_solvers = [minres]
```

So, I'm having some conflicts due to some of these solvers not even having the atol parameter to start with. I suppose this is the reason `_get_atol()` was implemented in the first place.